### PR TITLE
Manifold for `pipe_junction`.

### PR DESCRIPTION
**Pull request by marcfehling**
Follow-up to #13028. (EDIT: Has been merged.)

Adds the missing manifold description to the `pipe_junction` geometry.

### CODE REVIEW

**Comment by masterleinad** [comment]
Looks like the indentation is off:
```
diff --git a/source/grid/grid_generator_pipe_junction.cc b/source/grid/grid_generator_pipe_junction.cc
index de0c54f8..8f9937ab 100644
--- a/source/grid/grid_generator_pipe_junction.cc
+++ b/source/grid/grid_generator_pipe_junction.cc
@@ -384,8 +384,8 @@ namespace GridGenerator
     // Create a hyperball domain in 2D that will act as the reference cross
     // section for each pipe segment.
     const auto tria_base = []() {
-      constexpr unsigned int dim      = 3;
-      constexpr unsigned int spacedim = 3;
+      constexpr unsigned int               dim      = 3;
+      constexpr unsigned int               spacedim = 3;
       Triangulation<dim - 1, spacedim - 1> tria_base;
       GridGenerator::hyper_ball_balanced(tria_base,
                                          /*center=*/Point<spacedim - 1>(),
```

File: source/grid/grid_generator_pipe_junction.cc [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+          // Further, to be able to smoothly merge all pipe segments at the
+          // bifurcation, we also need to transform in z-direction.
+          const double z_factor =
+            PipeSegment::compute_z_expansion(x_new, y_new, data);
+          Assert(z_factor > 0,
+                 ExcMessage("Invalid input: at least one pipe segment "
+                            "is not long enough in this configuration"));
+          const double z_new = z_factor * pt[2];
+
+          return Point<spacedim>(x_new, y_new, z_new);
```
Comment (masterleinad) [inline source/grid/grid_generator_pipe_junction.cc:542]:
```
D:\a\dealii\dealii\source\grid\grid_generator_pipe_junction.cc(542,1): error C2975: 'dim': invalid template argument for 'dealii::Point', expected compile-time constant expression [D:\a\dealii\dealii\build\source\grid\obj_grid_debug.vcxproj]
D:\a\dealii\dealii\include\deal.II/base/point.h(109): message : see declaration of 'dim' [D:\a\dealii\dealii\build\source
```

**Comment by marcfehling** [comment]
Rebased and ready for review!

**Review by kronbichler — APPROVED** [review]

File: source/grid/grid_generator_pipe_junction.cc [hunk 2]
```diff
@@ -69,6 +69,194 @@ namespace
...
+       * Compute the Cartesian coordinates for a chart point given in
+       * cylindrical coordinates $(r, \phi, \lambda)$ on a cylinder of height
+       * one, where $r$ denotes the distance from the axis, $\phi$ the angle
+       * between the given point and the normal direction, and $\lambda$ the
+       * axial position.
+       */
+      virtual Point<spacedim>
+      push_forward(const Point<3> &chart_point) const override;
+
+    protected:
```
Comment (bangerth) [inline source/grid/grid_generator_pipe_junction.cc:135]: What is the reason why these members could not be `private` as well?

File: source/grid/grid_generator_pipe_junction.cc [hunk 3]
```diff
@@ -69,6 +69,194 @@ namespace
...
+    }
+
+
+
+    template <int dim, int spacedim>
+    std::unique_ptr<dealii::Manifold<dim, spacedim>>
+    Manifold<dim, spacedim>::clone() const
+    {
+      return std::make_unique<Manifold<dim, spacedim>>(
+        normal_direction, direction, point_on_axis, data, tolerance);
```
Comment (bangerth) [inline source/grid/grid_generator_pipe_junction.cc:206]:
```suggestion
      return std::make_unique<Manifold<dim, spacedim>>(*this);
```
:-)

File: source/grid/grid_generator_pipe_junction.cc [hunk 4]
```diff
@@ -69,6 +69,194 @@ namespace
...
+        Physics::VectorRelations::signed_angle(normal_direction,
+                                               p_diff,
+                                               /*axis=*/direction);
+
+      // Map the axial coordinate to a cylinder of height one.
+      lambda /= compute_z_expansion(r * std::cos(phi), r * std::sin(phi), data);
+
+      // Return distance from the axis, angle and signed distance on the axis.
+      return {r, phi, lambda};
+    }
```
Comment (bangerth) [inline source/grid/grid_generator_pipe_junction.cc:238]: This is in essence what the `Cylinder` manifold probably needs to do, except for an arbitrarily located axis, right?

**Review by bangerth — APPROVED** [review]
Let's get this merged!

File: source/grid/grid_generator_pipe_junction.cc (hunk 4 above)
Comment (marcfehling) [reply source/grid/grid_generator_pipe_junction.cc:238]: That is exactly what the `Cylinder` manifold does. Basically only line 234 is new. However, with this one additional line, I still thought that adding a dedicated `Manifold` class for this geometry is the way to go here.

File: source/grid/grid_generator_pipe_junction.cc (hunk 3 above)
Comment (marcfehling) [reply source/grid/grid_generator_pipe_junction.cc:206]: Ah, this is nice.

Might be a good idea to update the `clone` function from the other `Manifold` classes in `grid/manifold_lib.cc` as well.

File: source/grid/grid_generator_pipe_junction.cc (hunk 2 above)
Comment (marcfehling) [reply source/grid/grid_generator_pipe_junction.cc:135]: Honestly, I copied the class's structure from `CylindricalManifold`, which has them `protected` as well. I don't know why they were protected in the first class.

Comment (bangerth) [reply source/grid/grid_generator_pipe_junction.cc:135]: See https://github.com/dealii/dealii/pull/13307.

File: source/grid/grid_generator_pipe_junction.cc (hunk 3 above)
Comment (bangerth) [reply source/grid/grid_generator_pipe_junction.cc:206]: See https://github.com/dealii/dealii/pull/13307.

**Pull request merged**